File: typescript/src/search/binary.ts
/*
 * Binary search assumes that the input array is sorted
 */
export default function binarySearch(sortedArr: Array<number>, needle: number) {
  let low = 0;
  let high = sortedArr.length;

  while (low < high) {
    const middle = Math.floor(low + (high - low) / 2);

    let value = sortedArr[middle];

    if (value === needle) {
      return true;
    } else if (needle > value) {
      low = middle + 1;
    } else {
      high = middle;
    }
  }

  return false;
}

File: go/search/binary/binary.go
package binary

import "math"

func BinarySearch(haystack []int64, needle int64) bool {
	var low, high int64 = 0, int64(len(haystack))

	for low < high {
		mid := int64(math.Floor(float64(low + (high-low)/2)))

		val := haystack[mid]

		if val == needle {
			return true
		} else if needle > val {
			low = mid + 1
		} else {
			high = mid
		}

	}

	return false
}
